modules/extensions/src/terrain/terrain-effect.ts

```diff
@@ -0,0 +1,238 @@
+import {Texture2D, ProgramManager, Framebuffer} from '@luma.gl/core';
+import {log} from '@deck.gl/core';
+
+import {terrainModule, TerrainModuleSettings} from './shader-module';
+import {TerrainCover} from './terrain-cover';
+import {TerrainPass} from './terrain-pass';
+import {TerrainPickingPass, TerrainPickingPassRenderOptions} from './terrain-picking-pass';
+import {createRenderTarget} from './utils';
+
+import type {Effect, PreRenderOptions, Layer, Viewport} from '@deck.gl/core';
+
+/** Class to manage terrain effect */
+export class TerrainEffect implements Effect {
+  id = 'terrain-effect';
+  props = null;
+  useInPicking = true;
+
+  /** true if inside a picking pass */
+  private isPicking: boolean = false;
+  /** An empty texture as placeholder */
+  private dummyHeightMap: Texture2D;
+  /** A texture encoding the ground elevation, updated once per redraw. Used by layers with offset mode */
+  private heightMap?: Framebuffer;
+  private terrainPass!: TerrainPass;
+  private terrainPickingPass!: TerrainPickingPass;
+  /** One texture for each primitive terrain layer, into which the draped layers render */
+  private terrainCovers: Map<string, TerrainCover> = new Map();
+
+  initialize(gl: WebGLRenderingContext) {
+    this.dummyHeightMap = new Texture2D(gl, {
+      width: 1,
+      height: 1
+    });
+    this.terrainPass = new TerrainPass(gl, {id: 'terrain'});
+    this.terrainPickingPass = new TerrainPickingPass(gl, {id: 'terrain-picking'});
+
+    if (!this.heightMap) {
+      try {
```

Comment:
I would move the `if (!Framebuffer.isSupported(gl, {colorBufferFloat: true}))` check here. `createRenderTarget` may throw other, unrelated errors